staging/src/k8s.io/client-go/tools/cache/reflector_metrics.go

```diff
@@ -14,76 +14,45 @@ See the License for the specific language governing permissions and
+	"k8s.io/component-base/metrics/legacyregistry"
 )
 
-// GaugeMetric represents a single numerical value that can arbitrarily go up
-// and down.
-type GaugeMetric interface {
-	Set(float64)
-}
-
-// CounterMetric represents a single numerical value that only ever
-// goes up.
-type CounterMetric interface {
-	Inc()
-}
-
-// SummaryMetric captures individual observations.
-type SummaryMetric interface {
-	Observe(float64)
-}
-
-type noopMetric struct{}
-
-func (noopMetric) Inc()            {}
-func (noopMetric) Dec()            {}
-func (noopMetric) Observe(float64) {}
-func (noopMetric) Set(float64)     {}
-
-// MetricsProvider generates various metrics used by the reflector.
-type MetricsProvider interface {
-	NewListsMetric(name string) CounterMetric
-	NewListDurationMetric(name string) SummaryMetric
-	NewItemsInListMetric(name string) SummaryMetric
-
-	NewWatchesMetric(name string) CounterMetric
-	NewShortWatchesMetric(name string) CounterMetric
-	NewWatchDurationMetric(name string) SummaryMetric
-	NewItemsInWatchMetric(name string) SummaryMetric
-
-	NewLastResourceVersionMetric(name string) GaugeMetric
-}
-
-type noopMetricsProvider struct{}
+const (
+	reflectorSubsystem = "reflector"
+)
 
-func (noopMetricsProvider) NewListsMetric(name string) CounterMetric         { return noopMetric{} }
-func (noopMetricsProvider) NewListDurationMetric(name string) SummaryMetric  { return noopMetric{} }
-func (noopMetricsProvider) NewItemsInListMetric(name string) SummaryMetric   { return noopMetric{} }
-func (noopMetricsProvider) NewWatchesMetric(name string) CounterMetric       { return noopMetric{} }
-func (noopMetricsProvider) NewShortWatchesMetric(name string) CounterMetric  { return noopMetric{} }
-func (noopMetricsProvider) NewWatchDurationMetric(name string) SummaryMetric { return noopMetric{} }
-func (noopMetricsProvider) NewItemsInWatchMetric(name string) SummaryMetric  { return noopMetric{} }
-func (noopMetricsProvider) NewLastResourceVersionMetric(name string) GaugeMetric {
-	return noopMetric{}
-}
+var (
+	// reflectorListCount is the number of list requests issued by a reflector in order to start
+	// a watch broken down by the group-version-kind and the resource-version used.
+	reflectorListCount = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Subsystem:      reflectorSubsystem,
+			Name:           "watch_list_count",
+			Help:           "Number of list requests issued by reflector to start a watch",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{
+			"group_version_kind", // Gvk of the resouce the reflector is watching.
+			"resource_version",   // Resource version [empty, zero, explicit] used in the list request.
+		},
+	)
+)
 
-var metricsFactory = struct {
-	metricsProvider MetricsProvider
-	setProviders    sync.Once
-}{
-	metricsProvider: noopMetricsProvider{},
+func resourceVersionToLabelValue(rv string) string {
+	switch rv {
+	case "":
+		return "empty"
+	case "0":
+		return "zero"
+	default:
+		return "explicit"
+	}
 }
```

Comment:
What is the reasoning behind this?

Reply:
W want to keep the cardinality of the metric low. We don't really care about the exact value of the RV, but we want to distinguish these three different cases (rv="", rv=0, rv set explicitly) as they have different performance implications (either reading from etcd or watchcache) and can tell whether reflector is healthy or not.